packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -36,19 +37,94 @@ enum _ScaffoldSlot {
   statusBar,
 }
 
+/// A position that the [FloatingActionButton] can dock to in a [Scaffold].
+/// 
+/// [getOffset] 
+class FabPosition {
+  /// End-aligned FAB, floating at the bottom of the screen.
+  /// 
+  /// This is the default alignment of FABs in Material apps.
+  static const FabPosition endFloat = const _EndFloatFab();
+
+  /// Centered FAB, floating at the bottom of the screen.
+  static const FabPosition centerFloat = const _CenterFloatFab();
+
+  const FabPosition(this.getOffset);
+
+  /// Places the FAB with the layout parameters of the [Scaffold].
+  /// 
+  /// [scaffoldSize] is the size of the whole [Scaffold].
+  /// [fabSize] is the size of the [FloatingActionButton] to be placed.
+  /// [snackBarSize] is the size of the Scaffold's [SnackBar] (if available).
+  /// [bottomSheetSize] is the size of the [BottomSheet] (if available).
+  /// [textDirection] is the direction used by the Scaffold (either RTL or LTR).
+  /// [contentBottom] is the height from the top where the Scaffold's body ends.
+  /// [horizontalPadding] is padding the Scaffold would like the FAB to observe.
+  final Offset Function(Size scaffoldSize, Size fabSize, Size snackBarSize, Size bottomSheetSize, TextDirection textDirection, double contentBottom, double horizontalPadding) getOffset;
```

Comment:
I'm worried that this long list of parameters might grow and get confusing over time as the FAB positioning spec changes. In addition an app developer that wants custom FAB positioning might need to depend on a completely different set of parameters.

Would it be possible to abstract away the set of inputs that are used to determine the position?
e.g have this method take no parameters, and the different implementations would keep members that allow them to get the inputs they need.

I would imagine a ScaffoldFabPosition class that keeps a reference to the ScaffoldState, and queries it for the inputs it needs (or alternatively maybe have ScaffoldState implement FabPosition?)

Construction is going to be tricky though as you end up with a dependency loop.
Maybe we can make the return value a type that can either be an Offset, or say "defer to scaffold startegy".
Then the ScaffoldState could internally construct a private FabPosition, provide it with all the dependencies, and defer to it when the "main" FabPosition says "defer to scaffold".

Reply:
Another thought - I can imagine use cases where we animate the return values of this function.
e.g we want the FAB positioned relative to some other UI element that is currently being animated.
We might want the return type to be able to say whether the position is being animated or not (so we will not trigger another animation if it is already animation)